### PR TITLE
COR-436: redirect to record after submitting, refactoring

### PR DESCRIPTION
**Pull request by internetti**

### CODE REVIEW

File: frontend/apps/pwa/src/features/recorder/RecordEditorContainer.tsx [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+          recorderActions.selectRecord({
+            recordId: currentRecord.ownerId,
+          }),
+        );
+      }
+    } else {
+      const recordResponse = await dispatch(postRecord(recordsToPost)).unwrap();
+      dispatch(recordActions.addMany(recordResponse));
+      history.push(`/browse/records/${recordResponse[0].id}`);
+    }
```
Comment (internetti) [inline frontend/apps/pwa/src/features/recorder/RecordEditorContainer.tsx:150]: this is the most relevant change

File: frontend/apps/pwa/src/reducers/recorder.ts [hunk 2]
```diff
@@ -1,54 +1,68 @@
...
-import { recordGlobalSelectors } from '../../reducers/records';
-import client from '../../app/client';
+import { RootState } from '../app/store';
+import client from '../app/client';
 
-export interface FormValue {
+import { FormInterface, selectFormOrSubFormById, selectRootForm } from './form';
+import {recordGlobalSelectors} from "./records";
+
+export interface FormValue extends Omit<Record, 'databaseId'> {
```
Comment (internetti) [inline frontend/apps/pwa/src/reducers/recorder.ts:16]: this type felt redundant, opinions welcome ^^

**Comment by neb42** [comment]
When refactoring components to use the pattern in the style guide we should keep their files together in the directory.
In this pr you added `RecordEditorContainer` in `features/recorder`. I think this should instead be:

```
features/
  recorder/
    RecordEditor/
      index.ts
      RecordEditor.container.tsx
      RecordEditor.component.tsx
```

This way whoever uses `RecordEditor` can be thought of as a the public component with the container/component files being private. By having both the component and container in `features/recorder` it suggests that they are both used elsewhere in the application.

File: frontend/apps/pwa/src/features/browser/RecordField.tsx [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+import { Link } from 'react-router-dom';
+import format from 'date-fns/format';
+
+type RecordFieldProps = {
+  field: FieldDefinition;
+  value: any;
+  subRecords: Record[] | undefined;
+};
+export const RecordField: FC<RecordFieldProps> = (props) => {
+  const { field, value, subRecords } = props;
```
Comment (neb42) [inline frontend/apps/pwa/src/features/browser/RecordField.tsx:12]: I assume this came from copy/pasting, but we should destruct props like `= ({ field, value, subRecords }) => {

File: frontend/apps/pwa/src/features/browser/RecordField.tsx [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+  function renderField(f: FieldDefinition, v: any) {
```
Comment (neb42) [inline frontend/apps/pwa/src/features/browser/RecordField.tsx:14]: I think using arrow functions are typically better as they have block scoping, clearer currying syntax, and align with modern js practices better.

File: frontend/apps/pwa/src/reducers/recorder.ts (hunk 2 above)
Comment (neb42) [reply frontend/apps/pwa/src/reducers/recorder.ts:16]: I find this a little difficult to interpret as we're both extending and omitting. I think this is worth keeping if there is a strong relationship between Record and FormValue. ie. if we add a field to Record we will typically want it in FormValue.

File: frontend/apps/pwa/src/features/browser/RecordField.tsx (hunk 4 above)
Comment (internetti) [reply frontend/apps/pwa/src/features/browser/RecordField.tsx:14]: completely agree, copy paste issues ...

File: frontend/apps/pwa/src/components/FieldEditor/FieldEditor.tsx [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+  WeekFieldEditor,
+  SingleSelectFieldEditor,
+  SubFormFieldEditor,
+  ReferenceFieldEditor,
+} from '../formFields';
+
+export const FieldEditor: FC<FieldEditorProps> = (props) => {
+  const {
+    field: { fieldType },
+  } = props;
```
Comment (neb42) [inline frontend/apps/pwa/src/components/FieldEditor/FieldEditor.tsx:20]:
```
export const FieldEditor: FC<FieldEditorProps> = ({ field: { fieldType }) => {
```

Comment (neb42) [reply frontend/apps/pwa/src/components/FieldEditor/FieldEditor.tsx:20]: ah sorry you need the full props below

File: frontend/apps/pwa/src/components/formFields/MultiSelectFieldEditor.tsx [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+      <select
+        className="form-control bg-dark text-light border-secondary"
+        id={field.id}
+        value={value || []}
+        multiple
+        onChange={(event) => {
+          const { options } = event.target;
+          const selected = Object.entries(options).filter((o) => o[1].selected);
+          setValue(selected.map((s) => s[1].value));
+        }}
```
Comment (neb42) [inline frontend/apps/pwa/src/components/formFields/MultiSelectFieldEditor.tsx:27]: This should live in an `handleOnChange` function

File: frontend/apps/pwa/src/components/formFields/types.ts [hunk 7]
```diff
@@ -0,0 +1,12 @@
+import { FieldDefinition } from 'core-api-client';
+
+import { FormValue } from '../../reducers/recorder';
+
+export type FieldEditorProps = {
+  field: FieldDefinition;
+  value: string | string[] | null;
+  setValue: (value: string | string[] | null) => void;
```
Comment (neb42) [inline frontend/apps/pwa/src/components/formFields/types.ts:8]: I think a more common naming convention here is `onValueChange`. The component doesn't know that you're "setting" a value, it just knows that you're running something when the value changes.

**Comment by neb42** [comment]
In regard to `frontend/apps/pwa/src/components/formFields/helpers.tsx` I think helper files are generally pretty bad as they quickly grow out of control. In this case some functions are common across fields and some are field specific. As a quick fix I would make a directory, `frontend/apps/pwa/src/components/formFields/helpers`, and add each helper function as a file.

File: frontend/apps/pwa/src/features/recorder/RecordEditor.component.tsx [hunk 8]
```diff
@@ -0,0 +1,77 @@
+import React, { FC } from 'react';
+import { FieldDefinition, FieldValue } from 'core-api-client';
+
+import { FormValue } from '../../reducers/recorder';
+import { FieldEditor } from '../../components/FieldEditor/FieldEditor';
+
+export type RecordEditorProps = {
```
Comment (neb42) [inline frontend/apps/pwa/src/features/recorder/RecordEditor.component.tsx:7]: Does this need to be exported? If not I'd just call it `Props`

File: frontend/apps/pwa/src/features/recorder/RecordEditor.component.tsx [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+  fields: FieldDefinition[];
+  values: FieldValue[];
+  setValue: (key: string, value: any) => void;
+  selectSubRecord: (subRecordId: string) => void;
+  addSubRecord: (ownerFieldId: string) => void;
+  subRecords: { [key: string]: FormValue[] };
+  saveRecord: () => void;
```
[the remark this answers is not in the document]
Comment (neb42) [reply frontend/apps/pwa/src/features/recorder/RecordEditor.component.tsx:14]: And when using the component typically `handle<Action>` although this does differ on the use case.

eg. `<RecordEditorComponent onSaveRecord={handleSaveRecord} />`

File: frontend/apps/pwa/src/features/recorder/index.ts [hunk 10]
```diff
@@ -0,0 +1 @@
+export { RecordEditorContainer } from './RecordEditor.container';
```
Comment (neb42) [inline frontend/apps/pwa/src/features/recorder/index.ts:1]:
```
export { RecordEditorContainer as RecordEditor  } from './RecordEditor.container';
```
Components using RecordEditor don't care that it's a container. Exporting like this allows us to refactor it to not use a container in the future if we want.

File: frontend/apps/pwa/src/features/recorder/RecordEditor.component.tsx [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+              </div>
+            </div>
+          </div>
+        </div>
+      </div>
+    </div>
+  );
+};
+
+export default RecordEditorComponent;
```
Comment (neb42) [inline frontend/apps/pwa/src/features/recorder/RecordEditor.component.tsx:77]: I don't think this line is needed as we're doing a named export

Comment (internetti) [reply frontend/apps/pwa/src/features/recorder/RecordEditor.component.tsx:77]: ah forgot to delete it

**Pull request merged**